src/Storages/ObjectStorage/MergeTree/StorageObjectStorageMergeTreePartImporterSink.cpp

```diff
@@ -0,0 +1,79 @@
+#include <Storages/ObjectStorage/MergeTree/StorageObjectStorageMergeTreePartImporterSink.h>
+
+namespace DB
+{
+
+StorageObjectStorageMergeTreePartImporterSink::StorageObjectStorageMergeTreePartImporterSink(
+        const DataPartPtr & part_,
+        const std::string & path_,
+        const ObjectStoragePtr & object_storage_,
+        const ConfigurationPtr & configuration_,
+        const std::optional<FormatSettings> & format_settings_,
+        const Block & sample_block_,
+        const std::function<void(MergeTreePartImportStats)> & part_log_,
+        const ContextPtr & context_)
+        : SinkToStorage(sample_block_)
+        , object_storage(object_storage_)
+        , configuration(configuration_)
+        , format_settings(format_settings_)
+        , sample_block(sample_block_)
+        , context(context_)
+        , part_log(part_log_)
+{
+    stats.part = part_;
+    stats.file_path = path_;
+    sink = std::make_shared<StorageObjectStorageSink>(
+        stats.file_path,
+        object_storage,
+        configuration,
+        format_settings,
+        sample_block,
+        context);
+}
+
+String StorageObjectStorageMergeTreePartImporterSink::getName() const
+{
+    return "StorageObjectStorageMergeTreePartImporterSink";
+}
+
+void StorageObjectStorageMergeTreePartImporterSink::consume(Chunk & chunk)
+{
+    if (stats.status.code != 0)
+        return;
+
+    try
+    {
+        sink->consume(chunk);
+        stats.read_bytes += chunk.bytes();
+        stats.read_rows += chunk.getNumRows();
+
+        stats.status = ExecutionStatus(0, "Success");
+    } catch (...) {
+        stats.status = ExecutionStatus(-1, "Error importing part");
+        part_log(stats);
+    }
+}
+
+void StorageObjectStorageMergeTreePartImporterSink::onFinish()
+{
+    if (stats.status.code != 0)
+    {
+        sink->cancel();
+        return;
+    }
+
+    sink->onFinish();
+    if (const auto object_metadata = object_storage->tryGetObjectMetadata(stats.file_path))
+    {
+        stats.bytes_on_disk = object_metadata->size_bytes;
+    }
+    part_log(stats);
+}
+
+void StorageObjectStorageMergeTreePartImporterSink::onException(std::exception_ptr)
+{
+    /// we should not reach here
+    std::terminate();
```

Comment:
are you sure?

Reply:
Nope, just stubs for now.

Part of the logic in this class is very hackish to keep the exceptions contained so that a single pipeline failure does not cause all the other pipelines to abort.